toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -48,6 +52,167 @@ auto VariantMatch(V&& v, Fs&&... fs) -> decltype(auto) {
+// enough information to add Unicode handling without completely destroying the
+// relevant optimizations.
+static auto ScanForIdentifierPrefix(llvm::StringRef text) -> llvm::StringRef {
+  // A table of booleans that we can use to classify bytes as being valid
+  // identifier (or keyword) characters. This is used in the generic,
+  // non-vectorized fallback code to scan for length of an identifier.
+  constexpr std::array<bool, 256> IsIdentifierByteTable = []() constexpr {
+    std::array<bool, 256> table = {};
+    for (char c = '0'; c <= '9'; ++c) {
+      table[c] = true;
+    }
+    for (char c = 'A'; c <= 'Z'; ++c) {
+      table[c] = true;
+    }
+    for (char c = 'a'; c <= 'z'; ++c) {
+      table[c] = true;
+    }
+    table['_'] = true;
+    return table;
+  }();
+
+#if __x86_64__
+  // This code uses a scheme derived from the techniques in Geoff Langdale and
+  // Daniel Lemire's work on parsing JSON[1]. Specifically, that paper outlines
+  // a technique of using two 4-bit indexed in-register look-up tables (LUTs) to
+  // classify bytes in a branchless SIMD code sequence.
+  //
+  // [1]: https://arxiv.org/pdf/1902.08318.pdf
+  //
+  // The goal is to get a bit mask classifying different sets of bytes. For each
+  // input byte, we first test for a high bit indicating a UTF-8 encoded Unicode
+  // character. Otherwise, we want the mask bits to be set with the following
+  // logic derived by inspecting the high nibble and low nibble of the input:
+  // bit0 = 1 for `_`: high `0x5` and low `0xF`
+  // bit1 = 1 for `0-9`: high `0x3` and low `0x0` - `0x9`
+  // bit2 = 1 for `A-O` and `a-o`: high `0x4` or `0x6` and low `0x1` - `0xF`
+  // bit3 = 1 for `P-Z` and 'p-z': high `0x5` or `0x7` and low `0x0` - `0xA`
+  // bit4 = unused
+  // bit5 = unused
+  // bit6 = unused
+  // bit7 = unused
+  //
+  // No bits set means definitively non-ID ASCII character.
+  //
+  // bits 4-7 remain unused if we need to classify more characters.
+  const auto high_lut = _mm_setr_epi8(
+      /*0x0:*/ 0b0000'0000,
+      /*0x1:*/ 0b0000'0000,
+      /*0x2:*/ 0b0000'0000,
+      /*0x3:*/ 0b0000'0010,
+      /*0x4:*/ 0b0000'0100,
+      /*0x5:*/ 0b0000'1001,
+      /*0x6:*/ 0b0000'0100,
+      /*0x7:*/ 0b0000'1000,
+      /*0x8:*/ 0b0000'0000,
+      /*0x9:*/ 0b0000'0000,
+      /*0xA:*/ 0b0000'0000,
+      /*0xB:*/ 0b0000'0000,
+      /*0xC:*/ 0b0000'0000,
+      /*0xD:*/ 0b0000'0000,
+      /*0xE:*/ 0b0000'0000,
+      /*0xF:*/ 0b0000'0000);
+  const auto low_lut = _mm_setr_epi8(
+      /*0x0:*/ 0b0000'1010,
+      /*0x1:*/ 0b0000'1110,
+      /*0x2:*/ 0b0000'1110,
+      /*0x3:*/ 0b0000'1110,
+      /*0x4:*/ 0b0000'1110,
+      /*0x5:*/ 0b0000'1110,
+      /*0x6:*/ 0b0000'1110,
+      /*0x7:*/ 0b0000'1110,
+      /*0x8:*/ 0b0000'1110,
+      /*0x9:*/ 0b0000'1110,
+      /*0xA:*/ 0b0000'1100,
+      /*0xB:*/ 0b0000'0100,
+      /*0xC:*/ 0b0000'0100,
+      /*0xD:*/ 0b0000'0100,
+      /*0xE:*/ 0b0000'0100,
+      /*0xF:*/ 0b0000'0101);
```

Comment:
I looked into whether we could use a similar approach as for `IsIdentifierByteTable` here, to compute this table rather than hard-coding it:

```suggestion
  constexpr auto luts = []() constexpr {
    std::array<std::array<std::uint8_t, 16>, 2> luts = {};
    auto AddRange = [&](int bit, unsigned char min, unsigned char max) {
      CARBON_CHECK((min >> 4) == (max >> 4));
      luts[1][min >> 4] |= 1 << bit;
      for (unsigned char c = min; c <= max; ++c) {
        luts[0][c & 0xF] |= 1 << bit;
      }
    };
    // This code uses a scheme derived from the techniques in Geoff Langdale
    // and Daniel Lemire's work on parsing JSON[1]. Specifically, that paper
    // outlines a technique of using two 4-bit indexed in-register look-up
    // tables (LUTs) to classify bytes in a branchless SIMD code sequence.
    //
    // [1]: https://arxiv.org/pdf/1902.08318.pdf
    //
    // The goal is to get a bit mask classifying different sets of bytes.
    // For each input byte, we first test for a high bit indicating a UTF-8
    // encoded Unicode character. Otherwise, we want the mask bits to be set
    // with the following logic derived by inspecting the high nibble and
    // low nibble of the input:

    // bit0 = 1 for `_`: high `0x5` and low `0xF`
    AddRange(0, '_', '_');

    // bit1 = 1 for `0-9`: high `0x3` and low `0x0` - `0x9`
    AddRange(1, '0', '9');

    // bit2 = 1 for `A-O` and `a-o`: high `0x4` or `0x6` and low `0x1` - `0xF`
    AddRange(2, 'A', 'O');
    AddRange(2, 'a', 'o');

    // bit3 = 1 for `P-Z` and 'p-z': high `0x5` or `0x7` and low `0x0` - `0xA`
    AddRange(3, 'P', 'Z');
    AddRange(3, 'p', 'z');

    // bit4 = unused
    // bit5 = unused
    // bit6 = unused
    // bit7 = unused
    //
    // No bits set means definitively non-ID ASCII character.
    //
    // bits 4-7 remain unused if we need to classify more characters.
    return luts;
  }();
  auto MakeMask = [&]<std::size_t... I>(int which,
                                        std::index_sequence<I...>) {
    return _mm_setr_epi8(luts[which][I]...);
  };
  const auto high_lut = MakeMask(1, std::make_index_sequence<16>());
  const auto low_lut = MakeMask(0, std::make_index_sequence<16>());
```

But in the end I think it's actually a lot easier to understand as just a table of data :)

Reply:
Yeah, the table is really helpful for visualizing the result of all of this.